lib/workflow/stores/mongo.js

```diff
@@ -883,5 +883,82 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
         return waterline.graphobjects.findOne(query);
     };
 
+    exports.checkoutTimer = function(leaseToken, criteria, leaseDuration) {
+        var now = new Date();
+        var query =  {
+            $and: [
+                criteria,
+                { paused: false },
+                {
+                    leaseToken: null,
+                    $or: [
+                        { nextScheduled: { $lte: now } },
+                        { nextScheduled: null }
+                    ]
+                }
+            ]
+        };
+
+        var options = {
+            sort: {nextScheduled: 1},
+            new: true
+        };
+        
+        var update = {
+            $set: {
+                lastStarted: now,
+                leaseToken: leaseToken,
+                leaseExpires: new Date(now.valueOf() + leaseDuration)
+            }
+        };
+
+        return  waterline.workitems.findAndModifyMongo(query, {}, update, options);
+    };
+
+    exports.updatePollerStatus = function(id, obj) {
+        var query = {
+            _id: waterline.workitems.mongo.objectId(id)
+        };
+
+        var update = {
+            $set: {
+                nextScheduled: obj.nextScheduled,
+                lastFinished: obj.lastFinished,
+                leaseToken: null,
+                leaseExpires: null,
+                state: obj.state
+            }
+        };
+
+        var options = {
+            new: true
+        };
+
+        if( obj.status !== Constants.Task.States.Succeeded ) {
+            update.$inc = {failureCount: 1};
+        } else {
+            update.$set.failureCount = 0;
```

Comment:
@benbp This keeps to the original behavior of "# of failures since last success": https://github.com/RackHD/on-core/pull/180/files#diff-00c99b3c7b76756226357f509811e470L199.